test/sudo-code

```diff
@@ -0,0 +1,88 @@
+runICE()
+{
+  setupC5()
+  guessInitialCandidate()
+  learnInv()
+}
+
+setupC5()
+{
+  lowerInterval = 2
+  upperInterval = 2
+
+  for each predicate p:
+  {
+    for each variable v in p.args:
+    {
+      if v is INT or v is BOOL:
+      {
+        write v as an attribute into C5.names file
+        upInterval++
+      }
+      for each INT pair (a,b):
+      {
+        write a+b, a-b as attributes into C5.names file
+        upInterval += 2
+      }
+      else:
+      {
+        print 'DO NOT SUPPORT THIS TYPE!'
+      }
+    }
+    write lowerInterval and upperInterval into C5.intervals file
+    lowerInterval = upperInterval
+    upperInterval = lowerInterval
+  }  
+}
+
+guessInitialCandidate()
+{
+  for each predicate p:
+  {
+    cand = true
+    cand_model.addDef(p, cand)
+  }
+}
+
+learnInv()
+{
+  put all horn rules into worklist
+  while worklist is not empty:
+  {
+    pick out one rule r
+    head = r.head()
+    body = r.body()
+    vc = !head AND body
+    replace each predicate in vc with its candidate, get vc'
+    
+    isValid = validateRule(r)
+    if(isValid)
+    {
+      continue
+    }
+    else
+    {
+      get model m from solver
+      extract attributes value from m             //?
+      extract implication pairs like (dp1 -> dp2) //?
+      add attributes as counter examples into C5.data file
+      C5learn()
+      add dependent rules back to worklist, including itself
+    }
+  }
+}
```

Comment:
I need a concrete example to illustrate how to implement your questions. This is really easy. For each attribute, like a variable `x`, you find its name in the pre-state and use `model.eval` to find its value. Then you find its name in the post-state, and use `model.eval` again. The result is a pair, value of `x` before the rule and value of `x` after the rule. Do it for all variables and you have an implication pair.

This is easiest in flat encoding since every variable has both source and destination versions.

Note that to extend this with E- and C-examples you'll also need to form the horn system we described on skype (with all the rules at once) and solve it with some bound. If the solution is sat, you'll mine the samples from there.